src/opt/dom_info.cpp

```diff
@@ -0,0 +1,300 @@
+    {
+        const instruction* iter = ins2;
+        while(instr2_idom.find(iter) != instr2_idom.end())
+        {
+            if(ins1 == instr2_idom[iter])
+                return true;
+            iter = instr2_idom[iter];
+        }
+    }
+    return false;
+}
+
+// Query whether ins1 strictly post-dominates ins2.  ins1 strictly post-dominates
+// ins2 if ins1 post-dominates ins2 and ins1 is not ins2.
+//
+bool dom_info::strictly_post_dominates(const instruction* ins1, const instruction* ins2)
+{
+    if(ins1 != ins2)
+    {
+        const instruction* iter = ins2;
+        while(instr2_ipdom.find(iter) != instr2_ipdom.end())
+        {
+            if(ins1 == instr2_ipdom[iter])
+                return true;
+            iter = instr2_ipdom[iter];
+        }
+    }
+    return false;
+}
+
+//  Among p_ins's dominators, find ones that strictly dominates or post-dominators others.
+//
+void dom_info::find_dom_tree(
+    std::unordered_map<const instruction*, std::set<const instruction*>>& instr2_doms,
+    const instruction* p_ins,
+    std::unordered_map<const instruction*, const instruction*>& instr2_dom_tree,
+    std::unordered_map<const instruction*, const instruction*>& idom)
+{
+    for(auto& iter1 : instr2_doms[p_ins])
+    {
+        auto dom_check = [& dom_tree = idom, ins1 = iter1 ](const instruction* ins2)
+        {
+            if(ins1 == ins2)
+                return false;
+            const instruction* iter = ins2;
+            ;
+            while(dom_tree.find(iter) != dom_tree.end())
+            {
+                if(ins1 == dom_tree[iter])
+                    return true;
+                iter = dom_tree[iter];
+            }
+            return false;
+        };
+
+        // check whether iter1 strictly dominates or post-dominates any other notes in
+        // p_ins's dominators or post-dominators.
+        if(!std::any_of(instr2_doms[p_ins].begin(), instr2_doms[p_ins].end(), dom_check))
+        {
+            assert(instr2_dom_tree.find(p_ins) == instr2_dom_tree.end());
+            instr2_dom_tree[p_ins] = iter1;
+        }
+    }
+}
+
+//  Compute dominator or post-dominator.  Instructions that do not use
+//  streams are left out.
+//
+void dom_info::compute_dom(bool reversed)
+{
+    std::size_t num_of_instrs = p_program->size();
+    if(num_of_instrs == 0)
+        return;
+    std::unordered_map<const instruction*, std::set<const instruction*>> instr2_doms;
+    std::unordered_map<const instruction*, int> instr2_points;
+    int cur_points   = reversed ? num_of_instrs - 1 : 0;
+    bool seen_stream = false;
+    program_visitor vis{p_program, reversed};
+    std::unordered_map<const instruction*, const instruction*>& instr2_dom_tree =
+        (reversed ? instr2_ipdom : instr2_idom);
+    for(auto ins = vis.begin(), end = vis.end();; ins = vis.next(ins))
+    {
+        const instruction* p_ins = &(*ins);
+        instr2_points[p_ins]     = cur_points;
+        if(ins->get_stream() < 0)
```

Comment:
Shouldn't the stream be read from the `set_stream` operators since that is the actual stream being used?